javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.expression
+
+import arcs.core.data.expression.InferredType.Primitive.BigIntType
+import arcs.core.data.expression.InferredType.Primitive.BooleanType
+import arcs.core.data.expression.InferredType.Primitive.ByteType
+import arcs.core.data.expression.InferredType.Primitive.DoubleType
+import arcs.core.data.expression.InferredType.Primitive.FloatType
+import arcs.core.data.expression.InferredType.Primitive.IntType
+import arcs.core.data.expression.InferredType.Primitive.LongType
+import arcs.core.data.expression.InferredType.Primitive.NullType
+import arcs.core.data.expression.InferredType.Primitive.NumberType
+import arcs.core.data.expression.InferredType.Primitive.ShortType
+import arcs.core.data.expression.InferredType.Primitive.TextType
+import arcs.core.data.expression.InferredType.ScopeType
+import arcs.core.data.expression.InferredType.SeqType
+import arcs.core.data.expression.InferredType.UnionType
+import arcs.core.util.BigInt
+import kotlin.test.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+/** Tests for [TypeEvaluator]. */
+@RunWith(JUnit4::class)
+class TypeCheckerTest {
+  /**
+   * Checks that the resulting type of the expression is compatible with the expected type. For
+   * example, a byte is assignable to an int, but an int is not assignable to a byte. A union
+   * of A|B is assignable to A|B|C but not vice-versa.
+   */
+  fun checkAssignableTo(
+    expr: Expression<*>,
+    expectedType: InferredType,
+    scope: Expression.Scope = MapScope<InferredType>("root", mapOf()),
+    params: Expression.Scope = ParameterScope()
+  ) {
+    val inferredType = expr.accept(TypeEvaluator(params), scope)
+    assertTrue(expectedType.isAssignableFrom(inferredType), "$expectedType != $inferredType")
+  }
+
+  @Test
+  fun check_literals() {
+    checkAssignableTo(1.0.asExpr(), DoubleType)
```

Comment:
you could hypothetically have a bug in assignability checking that always return `true` and all tests would pass. It seems to me it would be better if we also had some `checkNotAssignableTo` mixed in, wdyt? Alternatively, maybe we just need separate tests for assignability logic.

Reply:
Yeah, i thought of that and tested it manually :) But I'll add negative tests too.